### DISCUSSION TITLE
Reference custom fields and custom validation

### DISCUSSION THREAD

**Discussion by jbethuel** [firecmsco/firecms#246] [Q&A] [answered]
Hi first of all, thank you for your work!

Is reference custom fields possible? For example:
I've tried setting component for **config.Field** but its not picking up the data from the reference
```
{
  "reference_fields": [
    {
       id: "ref-1",
    }
  ]
}

// references
[
  {
     id: "ref-1"
  },
  {
     id: "ref-2"
  }
]
```

For the custom validations:
```
phone: {
  title: "Phone",
  dataType: "string",
  validation: {
     fn: (value) => phonenumberUtils.isValid(value) // or an object from yup schema
  }
}
```

or would it be better to move it in saving and deleting callbacks?

Again, thanks for the awesome package! 🎉

**Comment by 5hee75** [comment]
I am also wondering how to customize how reference fields are stored.  For example, we de-normalize some of our data between documents and collections, so ideally we'd like to maintain that and just add a new `reference` dataType field to that de-normalized map.

For example, if we were tracking the `owner` of a document,  in the CMS interface it should look like a reference field and be editable like a reference field, but we would preserve the other data along with the "reference".

```js
// Without "reference" 
owner: {
  id:  'abcdefg',
  name: 'John Doe',
  age: 25,
  favoriteColor: 'blue',
}

// With "reference"
owner: {
  id:  'abcdefg',
  name: 'John Doe',
  age: 25,
  favoriteColor: 'blue',
  ref: '/users/abcdefg';  // New "reference" field used by CMS interface
}
```

Would it make sense to define the `owner` field as a "map" data type and somehow customize the `Field`, `Preview`, and `onPreSave` to handle the rest of the data?  Is there an easier way to do it?

**Comment by jbethuel** [comment, marked as answer]
Managed to get it working by creating **CustomField** using **useDataSource** hook to fetch other collection.
